### PR TITLE
Additional upgrade e2e tests

### PR DESCRIPTION
**Pull request by skriss**
**What this PR does / why we need it**: Add basic upgrade tests for DaemonSet and Job, and add "during upgrade" testing to ConfigMap test. Add a simple harness for testing upgrade tests.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:

**Special notes for your reviewer**: continuation of #41296 @krousey please review, thanks

**Release note**: `NONE`

### CODE REVIEW

**Comment by spxtr** [comment]
/approve
@k8s-bot ok to test

File: test/e2e/framework/util.go [hunk 1]
```diff
@@ -928,7 +928,7 @@ func deleteNS(c clientset.Interface, clientPool dynamic.ClientPool, namespace st
 	}
 
 	// wait for namespace to delete or timeout.
-	err := wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+	err := wait.PollImmediate(Poll, timeout, func() (bool, error) {
```
Comment (krousey) [inline test/e2e/framework/util.go]: Why is this change here? Was it needed or was this just cleanup?

File: test/e2e/upgrades/daemonsets.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+			nodeNames = append(nodeNames, node.Name)
+		} else {
+			framework.Logf("Node %v not expected to have DaemonSet pod, has taints %v", node.Name, taints)
+		}
+	}
+
+	return checkDaemonPodOnNodes(f, namespace, selector, nodeNames)
+}
+
+func checkDaemonPodOnNodes(f *framework.Framework, namespace string, labelSet map[string]string, nodeNames []string) (bool, error) {
```
Comment (krousey) [inline test/e2e/upgrades/daemonsets.go]: This function always returns a nil error.

File: test/e2e/upgrades/daemonsets.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+// Teardown cleans up any remaining resources.
+func (t *DaemonSetUpgradeTest) Teardown(f *framework.Framework) {
+	// rely on the namespace deletion to clean up everything
+}
+
+func (t *DaemonSetUpgradeTest) validateRunningDaemonSet(f *framework.Framework) {
+	// Pods should come up in a reasonable amount of time
+	By("confirming the DaemonSet pods are running on all expected nodes")
+	_, err := checkRunningOnAllNodes(f, t.daemonSet.Namespace, t.daemonSet.Labels)
+	framework.ExpectNoError(err)
```
Comment (krousey) [inline test/e2e/upgrades/daemonsets.go]: This error is always nil and the boolean that actually indicates if the pods are running is ignored.

File: test/e2e/upgrades/mock_cluster_upgrade.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+var upgradeTests = []Test{
+	&ServiceUpgradeTest{},
+	&SecretUpgradeTest{},
+	&DeploymentUpgradeTest{},
+	&ConfigMapUpgradeTest{},
+	&JobUpgradeTest{},
+	&DaemonSetUpgradeTest{},
+}
+
+var _ = framework.KubeDescribe("MockUpgrade [Feature:MockUpgrade]", func() {
```
Comment (krousey) [inline test/e2e/upgrades/mock_cluster_upgrade.go]: What purpose does this serve? It's easy enough to comment out the disruptive upgrade logic in the actual test.

File: test/e2e/upgrades/configmaps.go [hunk 5]
```diff
@@ -135,12 +155,33 @@ func (t *ConfigMapUpgradeTest) testPod(f *framework.Framework) {
...
-		"content of file \"/etc/configmap-volume/data\": some configmap data",
-		"mode of file \"/etc/configmap-volume/data\": -rw-r--r--",
+	pod, err := f.ClientSet.Core().Pods(t.configMap.Namespace).Create(pod)
+	framework.ExpectNoError(err)
+
+	return pod, err
+}
+
+// validate that the ConfigMap data is read correctly using both methods
+func (t *ConfigMapUpgradeTest) testPod(f *framework.Framework) {
```
Comment (krousey) [inline test/e2e/upgrades/configmaps.go]: This has fundamentally changed the test now, and I'm afraid it will no longer work during a cluster upgrade. It now relies on a pod you create in `Setup` instead of creating one itself. So during a cluster upgrade, the pod created in `Setup` will have its node upgraded and will no longer be running. Then this command will run after the upgrade and not be able to exec inside the pod.

This workflow is a little complicated to get right. I suggest either using a deployment to have it automatically rescheduled, or making sure a pod is running first, and if not, create it.

And since this is going to be reworked a little, could you do this change in a separate PR? I'd really like to get your daemon set test in. :smile:

**Review by krousey — CHANGES_REQUESTED** [review]
@kow3ns is already tackling Jobs in https://github.com/kubernetes/kubernetes/pull/41271

File: test/e2e/framework/util.go (hunk 1 above)
Comment (skriss) [inline test/e2e/framework/util.go]: This was just cleanup, it's non-essential; I noticed namespace deletion was a bit slow so was trying to tighten things up a bit. If this is risky I can undo.

File: test/e2e/upgrades/mock_cluster_upgrade.go (hunk 4 above)
Comment (skriss) [inline test/e2e/upgrades/mock_cluster_upgrade.go]: convenience only (so I don't have to think about commenting/uncommenting or accidentally checking in commented-out tests); if this kind of thing shouldn't be included, that's fine w.me.

**Comment by skriss** [comment]
@k8s-bot gci gce e2e test this

**Comment by skriss** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by skriss** [comment]
@k8s-bot gci gce e2e test this

**Comment by skriss** [comment]
@krousey this is ready for re-review. i pared it down to just the DaemonSet test and will submit ConfigMap changes in a separate PR.

File: test/e2e/upgrades/daemonsets.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+// Test validates that the DaemonSet is running during the upgrade (if applicable) and
+// verifies again post-upgrade that it's running
+func (t *DaemonSetUpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	testDuringDisruption := upgrade == MasterUpgrade
+
+	if testDuringDisruption {
+		By("validating the DaemonSet is still running during upgrade")
+
+		wait.Until(func() {
+			t.validateRunningDaemonSet(f)
```
Comment (krousey) [inline test/e2e/upgrades/daemonsets.go]: I'm not sure you can do this during a disruption. The master will be down at some point and your API calls will fail.

Other upgrade tests only do testing during disruption that don't involve calling the API. Like the service test  that accesses the service through a load balancer.

**Review by krousey — CHANGES_REQUESTED** [review]

File: test/e2e/framework/util.go [hunk 7]
```diff
@@ -3760,6 +3760,12 @@ func RunHostCmd(ns, name, cmd string) (string, error) {
 	return RunKubectl("exec", fmt.Sprintf("--namespace=%v", ns), name, "--", "/bin/sh", "-c", cmd)
 }
 
+// RunHostCmdOnContainer runs the given cmd in the context of the given pod/container
+// using `kubectl exec` inside of a shell.
+func RunHostCmdOnContainer(ns, podName, containerName, cmd string) (string, error) {
+	return RunKubectl("exec", fmt.Sprintf("--namespace=%v", ns), podName, fmt.Sprintf("--container=%v", containerName), "--", "/bin/sh", "-c", cmd)
+}
+
```
Comment (krousey) [inline test/e2e/framework/util.go]: Is this needed?

**Comment by krousey** [comment]
Just one last nit, and I think this is ready.

File: test/e2e/framework/util.go (hunk 7 above)
Comment (skriss) [inline test/e2e/framework/util.go]: argh. that was leftover from configmap work, thought for some reason it was being used in daemonset. will remove.

**Review by krousey — APPROVED** [review]

**Comment by krousey** [comment]
/lgtm

Thanks!

**Comment by krousey** [comment]
@k8s-bot cvm gce e2e test this

**Comment by skriss** [comment]
@k8s-bot cvm gce e2e test this

**Comment by skriss** [comment]
@krousey is this ok to merge?

**Comment by krousey** [comment]
@skriss Yes it is. We use an automated submitter that only submits when all tests are passing. Head has been failing off and on for quiet some time, so this is still stuck in the queue.

http://submit-queue.k8s.io/#/queue

**Comment by skriss** [comment]
@k8s-bot bazel test this

**Comment by skriss** [comment]
@k8s-bot cvm gce e2e test this

**Comment by skriss** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by skriss** [comment]
@k8s-bot kubemark e2e test this

**Comment by skriss** [comment]
@k8s-bot kops aws e2e test this

**Comment by krousey** [comment]
I0227 00:33:47.544] k8s.io/kubernetes/test/e2e/upgrades/daemonsets.go:128: undefined: "k8s.io/kubernetes/pkg/api/v1".GetNodeTaints
